internal/postgres/statement.go

```diff
@@ -0,0 +1,697 @@
+	//  Schemaname: schemas are PostgreSql namepaces; often unspecified; defaults to "public"
+	//  Relname: name of the table
+	// We build a table name from these three components as follows:
+	// a) nil components are dropped.
+	// b) if more than one component is specified, they are joined using "."
+	//    (Note that Spanner doesn't allow "." in table names, so this
+	//    will eventually get re-mapped when we construct the Spanner table name).
+	// c) Schemaname is dropped if it is "public".
+	// d) return error if Relname is nil or "".
+	var l []string
+	if n.Catalogname != nil {
+		l = append(l, *n.Catalogname)
+	}
+	if n.Schemaname != nil && *n.Schemaname != "public" { // Don't include "public".
+		l = append(l, *n.Schemaname)
+	}
+	if n.Relname == nil && *n.Relname == "" {
+		return "", "", fmt.Errorf("relname is empty: can't build table name")
+	}
+	l = append(l, *n.Relname)
+	pgTable := strings.Join(l, ".")
+	spTable, err := GetSpannerTable(conv, pgTable)
+	if err != nil {
+		return "", "", err
+	}
+	return pgTable, spTable, nil
+}
+
+type constraint struct {
+	ct   nodes.ConstrType
+	keys []string
+}
+
+// analyzeConstraints traverses a list of nodes (expecting them to be
+// Constraint nodes), and collects the contraints they represent as
+// a list of constraint-type/key-list pairs.
+func analyzeConstraints(conv *Conv, n nodes.Node, pgTable string, l []nodes.Node) (cs []constraint) {
+	for _, i := range l {
+		switch d := i.(type) {
+		case nodes.Constraint:
+			var keys []string
+			for _, j := range d.Keys.Items {
+				k, err := getString(j)
+				if err == nil {
+					keys = append(keys, k)
+				}
+				if err != nil {
+					conv.unexpected(fmt.Sprintf("Processing %v statement: error processing constraints: %w", reflect.TypeOf(n), err))
+					conv.errorInStatement([]nodes.Node{n, d})
+				}
+			}
+			cs = append(cs, constraint{d.Contype, keys})
+		default:
+			conv.unexpected(fmt.Sprintf("Processing %v statement: found %s node while processing constraints\n", reflect.TypeOf(n), reflect.TypeOf(d)))
+		}
+	}
+	return cs
+}
+
+// analyzeColDefConstraints is like analyzeConstraints, but is specifially for
+// ColDef constraints. These constraints don't specify a key since they
+// are constraints for the column defined by ColDef.
+func analyzeColDefConstraints(conv *Conv, n nodes.Node, pgTable string, l []nodes.Node, pgCol string) (cs []constraint) {
+	// Do generic constraint processing and then set the keys of each constraint
+	// to {pgCol}.
+	for _, c := range analyzeConstraints(conv, n, pgTable, l) {
+		if len(c.keys) != 0 {
+			conv.unexpected("ColumnDef constraint has keys")
+		}
+		c.keys = []string{pgCol}
+		cs = append(cs, c)
+	}
+	return cs
+}
+
+// getConstraint finds the constraint-type ct in the constraint list cs.
+// It returns true (along with the associated keys) if it finds ct,
+// and false otherwise.
+func getConstraint(ct nodes.ConstrType, cs []constraint) (bool, []string) {
```

Comment:
Perhaps switch the order of the results:
`([]string, bool)`
It's more consistent with the golang map lookup syntax of `val, ok := m['key']`

Reply:
Done.